File: 75-blind/linked-lists/linked-list-cycle.c
// Slow and fast pointers approach: time O(n), space O(1)

bool hasCycle(struct ListNode *head) {
    if(head == NULL) return false;
    struct ListNode* fast = head;
    struct ListNode* slow = head;
    while(fast->next != NULL && fast->next->next != NULL){
        fast = fast->next->next;
        slow = slow->next;
        if(fast == slow) return true;
    }
    return false;
}

File: 75-blind/linked-lists/linked-list-cycle.java
// Hashtable approach: time O(n), space O(n)

public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null) return false;
        HashSet<ListNode> visited = new HashSet<>();
        while(head.next != null){
            if(visited.contains(head)){
                return true;
            }
            visited.add(head);
            head = head.next;
        }
        return false;
    }
}
